### ISSUE TITLE
Podman doesn't connect behind corporate proxy

### ISSUE THREAD

**Issue by danny2001k** [containers/podman#8032]
/kind bug

**Description**

podman doesn't connect behind corporate proxy

**Steps to reproduce the issue:**
```
root@podmanhost01:/home/domain/user1# echo $http_proxy
http://domain\user1:userpass@proxy.company.com:8080
root@podmanhost01:/home/domain/user1# echo $https_proxy
http://domain\user1:userpass@proxy.company.com:8080
root@podmanhost01:/home/domain/user1# podman pull mariadb
```
**Describe the results you received:**
```
root@podmanhost01:/home/domain/user1# podman pull mariadb
Trying to pull docker.io/library/mariadb...
  Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp: lookup http: Temporary failure in name resolution
Trying to pull quay.io/mariadb...
  Get "https://quay.io/v2/": proxyconnect tcp: dial tcp: lookup http: Temporary failure in name resolution
Error: unable to pull mariadb: 2 errors occurred:
        * Error initializing source docker://mariadb:latest: error pinging docker registry registry-1.docker.io: Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp: lookup http: Temporary failure in name resolution
        * Error initializing source docker://quay.io/mariadb:latest: error pinging docker registry quay.io: Get "https://quay.io/v2/": proxyconnect tcp: dial tcp: lookup http: Temporary failure in name resolution
```
**Describe the results you expected:**

Not an error :)

**Additional information you deem important (e.g. issue happens only occasionally):**

Docker works:
```
root@podmanhost01:/home/domain/user1# docker pull mariadb
Using default tag: latest
latest: Pulling from library/mariadb
d72e567cc804: Pull complete
0f3630e5ff08: Pull complete
b6a83d81d1f4: Pull complete
4bf2111ecf0e: Pull complete
9572d64978a0: Pull complete
bcc9953bffb3: Pull complete
de429570dda5: Pull complete
3652bc6ea9f9: Pull complete
6e4bf87041c8: Pull complete
96e489d6af27: Pull complete
baaf018282fc: Pull complete
1cca1ca0b2da: Pull complete
Digest: sha256:3e6a64c089460fb3403b5d3458ace9710d0be94f3dfdaaaeb74945dbb8e5671c
Status: Downloaded newer image for mariadb:latest
docker.io/library/mariadb:latest
```
**Output of `podman version`:**
```
Version:      2.1.1
API Version:  2.0.0
Go Version:   go1.15.2
Built:        Thu Jan  1 02:00:00 1970
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.16.1
  cgroupManager: systemd
  cgroupVersion: v1
  conmon:
    package: 'conmon: /usr/libexec/podman/conmon'
    path: /usr/libexec/podman/conmon
    version: 'conmon version 2.0.20, commit: '
  cpus: 4
  distribution:
    distribution: ubuntu
    version: "18.04"
  eventLogger: journald
  hostname: podmanhost01
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 4.15.0-121-generic
  linkmode: dynamic
  memFree: 5217595392
  memTotal: 8352567296
  ociRuntime:
    name: runc
    package: 'containerd.io: /usr/bin/runc'
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc10
      commit: dc9208a3303feef5b3839f4323d9beb36df0a9dd
      spec: 1.0.1-dev
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 4124045312
  swapTotal: 4124045312
  uptime: 40m 29.33s
registries:
  search:
  - docker.io
  - quay.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 0
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 2.0.0
  Built: 0
  BuiltTime: Thu Jan  1 02:00:00 1970
  GitCommit: ""
  GoVersion: go1.15.2
  OsArch: linux/amd64
  Version: 2.1.1

```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
Listing... Done
podman/unknown,now 2.1.1~2 amd64 [installed]
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
HyperV VM

```
cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

**Comment by rhatdan** [comment]
@mtrmac ideas?

**Comment by rhatdan** [comment]
If this is true it means that their is a bug in golang itself, since we are using the golang bindings for connecting to web services, which are supposed to follow these flags.

@vrothberg WDYT?

**Comment by vrothberg** [comment]
A Go bug would surprise me as it would impact many people at once.

@danny2001k, could you share `/etc/containers/containers.conf` and try setting the `env=["http_proxy=..."]` in the `[engines]` table? Does that work? `containers.conf` is the only offender I can imagine in our code base at the moment.

**Comment by mtrmac** [comment]
TL;DR: “be liberal in what you accept” bites again, combined with a lazy implementation in the standard library.  The `\` must be percent-encoded; a Go standard library bug might be worth filing.

---

> ```
> root@podmanhost01:/home/domain/user1# echo $http_proxy
> http://domain\user1:userpass@proxy.company.com:8080
> root@podmanhost01:/home/domain/user1# echo $https_proxy
> http://domain\user1:userpass@proxy.company.com:8080
> root@podmanhost01:/home/domain/user1# podman pull mariadb
> ```

> proxyconnect tcp:

The configuration was recognized as a proxy configuration, the code is trying to contact the proxy.

> dial tcp: lookup http: Temporary failure in name resolution

The code has decided that the proxy hostname is `http`.

---

Note:
```go
func main() {
	in := `http://domain\user1:userpass@proxy.company.com:8080`
	url, err := url.Parse(in)
	fmt.Printf("%v\n", err)
}
```
results in a `parse "http://domain\\user1:userpass@proxy.company.com:8080": net/url: invalid userinfo` because `\` is not a valid character in that part of the URL.

https://golang.org/pkg/net/http/#ProxyFromEnvironment says
> The environment values may be either a complete URL or a "host[:port]", in which case the "http" scheme is assumed. An error is returned if the value is a different form.

but what the function _actually_ does is not parse a host[:port] but
```go
                // proxy was bogus. Try prepending "http://" to it and
                // see if that parses correctly. If not, we fall
                // through and complain about the original one.
                if proxyURL, err := url.Parse("http://" + proxy); err == nil {
```
and, well,
```go
func main() {
	in := `http://http://domain\user1:userpass@proxy.company.com:8080`
	url, err := url.Parse(in)
	fmt.Printf("%#v %#v\n", url, err)
}
```
results in
```
&url.URL{Scheme:"http", Opaque:"", User:(*url.Userinfo)(nil), Host:"http:", Path:"//domain\\user1:userpass@proxy.company.com:8080", RawPath:"//domain\\user1:userpass@proxy.company.com:8080", ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""} <nil>
```
which matches the observed error with `Host:"http:"`.

---

I didn’t look into why it works with Docker. As a general principle, I’m not a fan of individual applications overriding standard library behavior. Either change the configuration, or change the standard library; a single change will make many applications functional. Every single application fixing up the standard library behavior (and having to _know_ it needs fixing up) does not scale at all.

**Comment by agrandville** [comment]
Same problem, I can't pull registries behind corporate proxy.

```
[agrandville@sv1396 containers]$ sudo podman run hello-world
Trying to pull registry.access.redhat.com/hello-world...
  Get https://registry.access.redhat.com/v2/: proxyconnect tcp: EOF
Trying to pull registry.fedoraproject.org/hello-world...
  Get https://registry.fedoraproject.org/v2/: proxyconnect tcp: EOF
Trying to pull registry.centos.org/hello-world...
  Get https://registry.centos.org/v2/: proxyconnect tcp: EOF

Trying to pull docker.io/library/hello-world...
  Get https://registry-1.docker.io/v2/: proxyconnect tcp: EOF
Error: unable to pull hello-world: 4 errors occurred:
        * Error initializing source docker://registry.access.redhat.com/hello-world:latest: error pinging docker registry registry.access.redhat.com: Get https://registry.access.redhat.com/v2/: proxyconnect tcp: EOF
        * Error initializing source docker://registry.fedoraproject.org/hello-world:latest: error pinging docker registry registry.fedoraproject.org: Get https://registry.fedoraproject.org/v2/: proxyconnect tcp: EOF
        * Error initializing source docker://registry.centos.org/hello-world:latest: error pinging docker registry registry.centos.org: Get https://registry.centos.org/v2/: proxyconnect tcp: EOF
        * Error initializing source docker://hello-world:latest: error pinging docker registry registry-1.docker.io: Get https://registry-1.docker.io/v2/: proxyconnect tcp: EOF

```

A tcpdump shows the problem

CUrl gracefull proxy request with CONNECT

```
1	17:20:38,415729	192.168.101.159	192.168.69.12	TCP	66	38130 → 3128 [SYN] Seq=0 Win=29200 Len=0 MSS=1460 SACK_PERM=1 WS=128
2	17:20:38,416535	192.168.69.12	192.168.101.159	TCP	62	3128 → 38130 [SYN, ACK] Seq=0 Ack=1 Win=65535 Len=0 MSS=1460 SACK_PERM=1
3	17:20:38,416551	192.168.101.159	192.168.69.12	TCP	54	38130 → 3128 [ACK] Seq=1 Ack=1 Win=29200 Len=0
4	17:20:38,416630	192.168.101.159	192.168.69.12	HTTP	198	CONNECT registry.access.redhat.com:443 HTTP/1.1 
5	17:20:38,455361	192.168.69.12	192.168.101.159	HTTP	93	HTTP/1.1 200 Connection established 
6	17:20:38,455374	192.168.101.159	192.168.69.12	TCP	54	38130 → 3128 [ACK] Seq=145 Ack=40 Win=29200 Len=0
7	17:20:38,612624	192.168.101.159	192.168.69.12	TLSv1.2	262	Client Hello
8	17:20:38,665385	192.168.69.12	192.168.101.159	TLSv1.2	1414	Server Hello
9	17:20:38,665412	192.168.101.159	192.168.69.12	TCP	54	38130 → 3128 [ACK] Seq=353 Ack=1400 Win=31280 Len=0
10	17:20:38,665423	192.168.69.12	192.168.101.159	TCP	154	3128 → 38130 [PSH, ACK] Seq=1400 Ack=353 Win=65535 Len=100 [TCP segment of a reassembled PDU]
...
```

podman without CONNECT 

```
8	16:40:39,995732	192.168.101.159	192.168.69.12	TCP	66	38084 → 3128 [SYN] Seq=0 Win=29200 Len=0 MSS=1460 SACK_PERM=1 WS=128
10	16:40:39,996939	192.168.69.12	192.168.101.159	TCP	62	3128 → 38084 [SYN, ACK] Seq=0 Ack=1 Win=65535 Len=0 MSS=1460 SACK_PERM=1
11	16:40:39,996953	192.168.101.159	192.168.69.12	TCP	54	38084 → 3128 [ACK] Seq=1 Ack=1 Win=29200 Len=0
12	16:40:39,997056	192.168.101.159	192.168.69.12	TLSv1	250	Client Hello
13	16:40:39,997929	192.168.69.12	192.168.101.159	TCP	60	3128 → 38084 [FIN, ACK] Seq=1 Ack=197 Win=65535 Len=0
14	16:40:39,997975	192.168.101.159	192.168.69.12	TCP	54	38084 → 3128 [FIN, ACK] Seq=197 Ack=2 Win=29200 Len=0
15	16:40:39,998897	192.168.69.12	192.168.101.159	TCP	60	3128 → 38084 [ACK] Seq=2 Ack=198 Win=65534 Len=0
EOF
```

**Comment by mtrmac** [comment]
@agrandville That’s certainly a different root cause (no `dial tcp: lookup http`), please file a separate report.

**Comment by rhatdan** [comment]
Since this is in the standard library, it is not something we can fix.  Should we close this?

**Comment by rhatdan** [comment]
@agrandville @mtrmac @vrothberg Where do we stand on this one?

**Comment by mtrmac** [comment]
AFAICT the original is user error + missing syntax checking in the standard library.

**Comment by rhatdan** [comment]
Therefore this is not something Podman can fix, Closing.  Reopen if I am mistaken.

**Comment by mdavis-xyz** [comment]
I've got the same issue, except my proxy does not require a username and password.
I'm trying to follow the comments here. e.g. look at `/etc/containers/containers.conf`. But there is no `/etc/containers` directory on my machine.

**Comment by rhatdan** [comment]
You need to create it and then podman will read the containers.conf file from there.

**Comment by mtrmac** [comment]
@mdavis-xyz Please file new reports instead of piling on to closed ones if they don’t actually match the situation. It’s just going to be lost here.
